internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+			if !ok {
+				leftOpen = false
+				if leftSet.Size() == 0 {
+					return res, nil
+				}
+				break
+			}
+			if msg.err != nil {
+				return nil, msg.err
+			}
+			for {
+				t, err := msg.iter.Next(ctx)
+				if err != nil {
+					msg.iter.Stop()
+					if storage.IterIsDoneOrCancelled(err) {
+						break
+					}
+					return nil, err
+				}
+				if processUsersetMessage(t.GetObject(), leftSet, rightSet) {
+					msg.iter.Stop()
+					res.Allowed = true
+					return res, nil
+				}
+			}
+		case msg, ok := <-rightChan:
+			if !ok {
+				rightOpen = false
+				break
+			}
+			if msg.err != nil {
+				return nil, msg.err
+			}
+			if processUsersetMessage(msg.userset, rightSet, leftSet) {
+				res.Allowed = true
+				return res, nil
+			}
+		}
+	}
+	return res, nil
+}
+
+func (c *LocalChecker) checkTTUFastPathV2(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset, iter storage.TupleKeyIterator) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkTTUFastPathV2")
+	defer span.End()
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	objectType := tuple.GetType(req.GetTupleKey().GetObject())
+	tuplesetRelation := rewrite.GetTupleToUserset().GetTupleset().GetRelation()
+	computedRelation := rewrite.GetTupleToUserset().GetComputedUserset().GetRelation()
+
+	possibleParents, err := typesys.GetDirectlyRelatedUserTypes(objectType, tuplesetRelation)
+	if err != nil {
+		return nil, err
+	}
+
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	leftChans := make([]chan *iteratorMsg, 0)
+	for _, parentType := range possibleParents {
+		r := req.clone()
+		r.TupleKey = &openfgav1.TupleKey{
+			Object:   tuple.BuildObject(parentType.GetType(), "ignore"),
+			Relation: computedRelation,
+			User:     r.GetTupleKey().GetUser(),
+		}
+		rel, err := typesys.GetRelation(parentType.GetType(), computedRelation)
+		if err != nil {
+			// NOTE: is there a better way to check and filter rather than skipping?
+			// other paths can be reachable
+			continue
+		}
+		leftChan, err := c.fastPathRewrite(cancellableCtx, r, rel.GetRewrite())
+		if err != nil {
+			return nil, err
+		}
+		leftChans = append(leftChans, leftChan)
+	}
+
+	if len(leftChans) == 0 {
+		// NOTE: this should be an error right?
+		return &ResolveCheckResponse{
+			Allowed: false,
+		}, nil
+	}
+
+	return c.resolveFastPath(ctx, leftChans, wrapIterator(TTUKind, iter))
+}
+
+// NOTE: Can we make this generic and move it to concurrency pkg?
+func fanInIteratorChannels(ctx context.Context, chans []chan *iteratorMsg) chan *iteratorMsg {
+	limit := len(chans)
+	pool := concurrency.NewPool(ctx, limit)
+	out := make(chan *iteratorMsg, limit)
+
+	for _, c := range chans {
+		pool.Go(func(ctx context.Context) error {
+			for v := range c {
+				if !concurrency.TrySendThroughChannel(ctx, v, out) {
+					if v.iter != nil {
+						v.iter.Stop()
+					}
+				}
+			}
+			return nil
+		})
+	}
+
+	go func() {
+		_ = pool.Wait()
+		close(out)
+	}()
```

Comment:
I'm confused why you wrapped this within a gorotuine. In other places in the code, we put the `pool.Wait`s inside `defer`, why is this different?

Plus, nobody is waiting on this goroutine?

Reply:
> Plus, nobody is waiting on this goroutine?

The listener of the close event is.

> pool.Waits inside defer, why is this different?

The defer here wouldn't make any sense as it would `block`. So it would have to be somewhere else. But I don't see the point of leaking that if we have the `close` signal to subscribe to.

Reply:
Ah makes sense, could we add a note in the code for this? 

```
go func() {
		_ = pool.Wait()
		// NOTE: the consumer of this channel will block waiting for it to close
		close(out)
	}()
```